bundles-lib/src/main/java/org/apache/metron/bundles/BundleClassLoaders.java

```diff
@@ -0,0 +1,353 @@
+ */
+public final class BundleClassLoaders {
+
+  private static volatile BundleClassLoaders bundleClassLoaders;
+  private volatile InitContext initContext;
+  private static final Logger logger = LoggerFactory.getLogger(BundleClassLoaders.class);
+
+  private final static class InitContext {
+
+    private final List<FileObject> extensionDirs;
+    private final Map<String, Bundle> bundles;
+
+    private InitContext(
+        final List<FileObject> extensionDirs,
+        final Map<String, Bundle> bundles) {
+      this.extensionDirs = extensionDirs;
+      this.bundles = bundles;
+    }
+  }
+
+  private BundleClassLoaders() {
+  }
+
+  /**
+   * @return The singleton instance of the BundleClassLoaders
+   */
+  public static BundleClassLoaders getInstance() {
+    BundleClassLoaders result = bundleClassLoaders;
+    if (result == null) {
+      synchronized (BundleClassLoaders.class) {
+        result = bundleClassLoaders;
+        if (result == null) {
+          bundleClassLoaders = result = new BundleClassLoaders();
+        }
+      }
+    }
+    return result;
+  }
+
+  public static void reset() {
+    getInstance().unInit();
+  }
+
+  private void unInit() {
+    synchronized (this) {
+      initContext = null;
+    }
+  }
+
+  /**
+   * Initializes and loads the BundleClassLoaders. This method must be called before the rest of the
+   * methods to access the classloaders are called and it can be safely called any number of times
+   * provided the same framework and extension working dirs are used.
+   *
+   * @param fileSystemManager the FileSystemManager
+   * @param extensionsDirs where to find extension artifacts
+   * @param props BundleProperties
+   * @throws FileSystemException if any issue occurs while exploding bundle working directories.
+   * @throws java.lang.ClassNotFoundException if unable to load class definition
+   * @throws IllegalStateException already initialized with a given pair of directories cannot
+   * reinitialize or use a different pair of directories.
+   */
+  public void init(final FileSystemManager fileSystemManager, final List<FileObject> extensionsDirs,
+      BundleProperties props)
+      throws FileSystemException, ClassNotFoundException, URISyntaxException {
+    if (extensionsDirs == null || fileSystemManager == null) {
+      throw new NullPointerException("cannot have empty arguments");
+    }
+    InitContext ic = initContext;
+    if (ic == null) {
+      synchronized (this) {
+        ic = initContext;
+        if (ic == null) {
+          initContext = ic = load(fileSystemManager, extensionsDirs, props);
+        }
+      }
+    } else {
+      boolean matching = CollectionUtils
+          .isEqualCollection(initContext.extensionDirs, extensionsDirs);
+      if (!matching) {
+        throw new IllegalStateException(
+            "Cannot reinitialize and extension/framework directories cannot change");
+      }
+    }
+  }
+
+  /**
+   * Should be called at most once.
```

Comment:
I do not have an explicit test, but I have to use this in the integration tests for example when run in maven.  I will write a test